price-feeder/oracle/provider/websocket_controller.go

```diff
@@ -0,0 +1,189 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/rs/zerolog"
+)
+
+type (
+	MessageHandler func(int, []byte)
+
+	// WebsocketController defines a provider agnostic websocket handler
+	// that manages reconnecting, subscribing, and receiving messages
+	WebsocketController struct {
+		ctx              context.Context
+		providerName     Name
+		websocketURL     url.URL
+		subscriptionMsgs []interface{}
+		messageHandler   MessageHandler
+		logger           zerolog.Logger
+		client           *websocket.Conn
+		mu               sync.Mutex
+	}
+)
+
+func NewWebsocketController(
+	ctx context.Context,
+	providerName Name,
+	websocketURL url.URL,
+	subscriptionMsgs []interface{},
+	messageHandler MessageHandler,
+	logger zerolog.Logger,
+) *WebsocketController {
+	return &WebsocketController{
+		ctx:              ctx,
+		providerName:     providerName,
+		websocketURL:     websocketURL,
+		subscriptionMsgs: subscriptionMsgs,
+		messageHandler:   messageHandler,
+		logger:           logger,
+	}
+}
+
+// Start will contniously loop and attempt connecting to the websocket
+// until a successsful connection is made. It then starts the ping
+// service and read listener in new go routines and sends subscription
+// messages  using the passed in subscription messages
+func (c *WebsocketController) Start() {
+	connectTicker := time.NewTicker(defaultReconnectTime)
+	defer connectTicker.Stop()
+
+	for {
+		err := c.connect()
+		if err == nil {
+			go c.readWebSocket()
+			go c.ping()
+			err = c.subscribe()
+			if err == nil {
+				return
+			}
+			c.close()
+		}
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-connectTicker.C:
+			continue
+		}
+	}
```

Comment:
Just another suggestion, for me it feels weird to check for `err == nil` :sweat_smile: 

```suggestion
	for {
		if err := c.connect(); err != nil {
			c.logger.Err(err).Msg("connect")
			select {
			case <-c.ctx.Done():
				return
			case <-connectTicker.C:
				continue
			}
		}

		go c.readWebSocket()
		go c.ping()

		if err := c.subscribe(); err != nil {
			c.logger.Err(err).Msg("subscribe")
			c.close()
		}
		return
	}
```